### PR TITLE
🚀 Major Enhancements to Voice Notes Sync

### PR DESCRIPTION
**Pull request by hjamet**
# 🚀 Major Enhancements to Voice Notes Sync

This pull request addresses the following issues:
- 🔗 https://github.com/mysticcoders/voicenotes-sync/issues/34
- 🔗 https://github.com/mysticcoders/voicenotes-sync/issues/32
- 🔗 https://github.com/mysticcoders/voicenotes-sync/issues/31

### 1. 📁 Sub-notes Support
- Sub-notes are now synchronized and listed in a subsection of the parent note's "Related Notes" section.

### 2. 🎨 Custom Instructions Support
- Custom instructions created by templates are added to the note in a "Custom" section.

### 3. 🧹 Tidy Transcript
- Tidy transcripts are now added normally to notes.
- A new option (default: True) has been added to replace original transcripts with their tidy version if available.

### 4. 📊 Summary and Key Points Prioritization
- If available, the note's summary and/or key points are displayed first, before the transcript, for easier reading.

### 5. ⏱️ Note Duration and Timestamp Correction
- Note properties are now in a more human-readable format.

## 💖 Thank You Note

From the bottom of my heart, thank you for this amazing plugin! I hope this contribution will bring joy to everyone! 🍡

### CODE REVIEW

**Comment by yshhwy** [comment]
Thank you for your contribution to the project! I tested the code, and all functionalities work as expected. The improvements are excellent and greatly enhance the project. Your efforts in addressing the issues and improving the code are greatly appreciated. Your hard work and dedication help make this project better for everyone. Looking forward to more of your valuable contributions in the future!

**Review by kinabalu — CHANGES_REQUESTED** [review]
can you merge from latest main, i've added eslint and prettier as a lot of the "updates" here are only cosmetic so makes it hard to read the PR for actual changes.

Also what I've been thinking about is for the processing of a "note" to be put into an object, and the building of the template itself to be a configurable setting with a sensible default in place that can always be reverted to. I'd like this a bit more I think

Might help with the tidy option that I don't see in the UI? What am I missing?

**Comment by hjamet** [comment]
I'll take care of any conflicts :)
As for the Tidy option, it replaces the original transcript with its "cleaned" version if generated :)

**Comment by hjamet** [comment]
- ✅ **Merged with the latest main branch** 🔄
- 🖌️ **Implementation of improvements:**
  - Integration of Jinja formatting for notes 📝
  - Update of date formats 📅
- 🐛 **Fixes:**
  - Resolution of typing issues 🧮
  - Package version updates 📦
- 🆕 **Additions:**
  - Inclusion of all missing fields to the Jinja template (creations, related and joined notes, tags and attachments) 🧩

File: types.ts [hunk 1]
```diff
@@ -118,3 +49,19 @@ export interface User {
...
+/* -------------------------------------------------------------------------- */
+/*                               VOICENOTES API                               */
+/* -------------------------------------------------------------------------- */
+
+export interface VoiceNoteSignedUrl {
+  url: string;
+}
+
+export interface VoiceNoteRecordings {
+  data: any[]; // Le type exact des éléments du tableau dépendra de la structure de vos enregistrements
```
Comment (kinabalu) [inline types.ts:62]: can we remove the comments here or write them in english?

File: main.ts [hunk 2]
```diff
@@ -1,301 +1,322 @@
...
-  prependDateFormat: 'YYYY-MM-DD',
+    automaticSync: true,
+    syncTimeout: 60,
+    downloadAudio: false,
+    syncDirectory: 'voicenotes',
+    deleteSynced: false,
+    reallyDeleteSynced: false,
+    todoTag: '',
+    prependDateFormat: 'YYYY-MM-DD',
+    noteTemplate: `# {{ title }}\n\nDate: {{ date }}\n{% if summary %}## Summary\n\n{{ summary }}{% endif %}\n{% if points %}## Main points\n\n{{ points }}{% endif %}\n{% if attachments %}## Attachments\n\n{{ attachments }}{% endif %}\n## Transcript\n\n{{ transcript }}\n{% if audio_link %}[Audio]({{ audio_link }}){% endif %}\n{% if todo %}## Todos\n\n{{ todo }}{% endif %}\n{% if email %}## Email\n\n{{ email }}{% endif %}\n{% if custom %}## Others\n\n{{ custom }}{% endif %}\n{% if tags %}## Tags\n\n{{ tags }}{% endif %}\n{% if related_notes %}# Related Notes\n\n{{ related_notes }}{% endif %}\n{% if subnotes %}## Subnotes\n\n{{ subnotes }}{% endif %}`,
```
Comment (kinabalu) [inline main.ts:20]: lets make this a multiline string so its easier to read and edit later

File: settings.ts [hunk 3]
```diff
@@ -191,41 +191,81 @@ export class VoiceNotesSettingTab extends PluginSettingTab {
...
 
-          await this.plugin.saveSettings();
-          await this.display();
-        })
+    new Setting(containerEl)
+      .setName('Custom Note Template')
+      .setDesc('Custom template for synced notes. Available variables: {{title}}, {{date}}, {{transcript}}, {{audio_link}}, {{summary}}, {{tidy}}, {{points}}, {{todo}}, {{email}}, {{custom}}')
+      .addTextArea((text) =>
+        text
+          .setPlaceholder('# {{ title }}\n\nDate: {{ date }}\n{% if summary %}## Summary\n\n{{ summary }}{% endif %}\n{% if points %}## Main points\n\n{{ points }}{% endif %}\n{% if attachments %}## Attachments\n\n{{ attachments }}{% endif %}\n## Transcript\n\n{{ transcript }}\n{% if audio_link %}[Audio]({{ audio_link }}){% endif %}\n{% if todo %}## Todos\n\n{{ todo }}{% endif %}\n{% if email %}## Email\n\n{{ email }}{% endif %}\n{% if custom %}## Others\n\n{{ custom }}{% endif %}\n{% if tags %}## Tags\n\n{{ tags }}{% endif %}\n{% if related_notes %}# Related Notes\n\n{{ related_notes }}{% endif %}\n{% if subnotes %}## Subnotes\n\n{{ subnotes }}{% endif %}')
```
Comment (kinabalu) [inline settings.ts:250]: lets just refer again to the noteTemplate here rather than duplicate the text.

File: main.ts [hunk 4]
```diff
@@ -1,301 +1,322 @@
...
+            attachments: attachments,
+        };
+
+        // Render the template using Jinja
+        let note = jinja.render(this.settings.noteTemplate, context);
+
+        // Add metadata
+        const metadata = `---
+    recording_id: ${recording.recording_id}
+    duration: ${Math.floor(recording.duration / 60000) > 0
```
Comment (kinabalu) [inline main.ts:239]: let's not do inline calculations like this, prefer externalizing into a function

File: main.ts [hunk 5]
```diff
@@ -1,301 +1,322 @@
...
+                ? `${Math.floor(recording.duration / 60000)}m`
+                : ''
+            }${Math.floor((recording.duration % 60000) / 1000).toString().padStart(2, '0')}s
+    created_at: ${recording.created_at}
+    updated_at: ${recording.updated_at}
+    ${recording.tags && recording.tags.length > 0 ? `tags: ${recording.tags.map((tag: { name: string }) => tag.name).join(',')}` : ''}
```
Comment (kinabalu) [inline main.ts:245]: same here, externalized this to a function

**Review by kinabalu — CHANGES_REQUESTED** [review]
getting closer -- lets clean up and make the changes and perhaps we can get a group of folks to help test the changes before we push.

my thought is we'll deploy this as a beta using BRAT and if we don't get any bugs out of it, push it live.

thank you for the code!

**Comment by hjamet** [comment]
Thank you from the bottom of my heart! 🦦 
I've just made the changes 🙃

**Review by kinabalu — APPROVED** [review]

**Comment by giittyy** [comment]
Great work! This is an awesome update. It seems to be pulling dates correctly base on local timezone for the title/document portion. Can we please also have this for the created_at and updated_at properties?

**Comment by hjamet** [comment]
Thank you! I'll check it out if I have a moment ;)

**Pull request merged**